### PR TITLE
Relax tolerances in convolution link tests when using old cuDNN

### PR DESCRIPTION
**Pull request by emcastillo**
Refactor links_tests convolution nd and 2d to use the backend injection and testing.LinksTestCase
format.

Merge before #7864

### CODE REVIEW

File: tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py [hunk 1]
```diff
@@ -5,137 +5,106 @@
...
-                                       (2, 2, 2, 2)).astype(self.x_dtype)
-        self.check_backward_options = {}
+        self.check_backward_options.update({
+            'atol': 5e-4, 'rtol': 5e-3
+        })
         if self.x_dtype == numpy.float16 or self.W_dtype == numpy.float16:
-            self.check_backward_options = {'atol': 3e-2, 'rtol': 5e-2}
+            self.check_forward_options.update({'atol': 5e-3, 'rtol': 5e-2})
+            self.check_backward_options.update({
+                'atol': 1e-3, 'rtol': 1e-3
```
Comment (niboshi) [inline tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py]: Tolerances are modified. Is it intentional?

Comment (emcastillo) [reply tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py]: Yes, original test had retries and it fails a lot when they are removed.

File: tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py [hunk 2]
```diff
@@ -5,137 +5,101 @@
...
+                                 (2, 3, 4, 3)).astype(self.x_dtype)
+        return x,
+
+    def forward_expected(self, link, inputs):
+        x, = inputs
+        y = link(x).array
+        return y,
+
+    def test_pickling(self, backend_config):
+        with backend_config:
```
Comment (niboshi) [inline tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py]: Shouldn't this test succeed without `with backend_config`?

File: tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py [hunk 3]
```diff
@@ -151,74 +115,77 @@ def test_col2im_consistency(self):
...
+        x, = inputs
+        y = link(x).array
+        return y,
 
-    def check_pickling(self, x_data):
-        x = chainer.Variable(x_data)
-        y = self.link(x)
-        y_data1 = y.data
+    def test_pickling(self, backend_config):
+        with backend_config:
```
Comment (niboshi) [inline tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py]: Ditto.

File: tests/chainer_tests/links_tests/connection_tests/test_convolution_nd.py [hunk 4]
```diff
@@ -26,37 +23,112 @@
...
+        x = numpy.random.uniform(-1, 1, self.x_shape).astype(self.dtype)
+        return x,
+
+    def forward_expected(self, link, inputs):
+        x, = inputs
+        y = link(x).array
+        return y.astype(self.dtype),
+
+    def test_pickling(self, backend_config):
+        with backend_config:
```
Comment (niboshi) [inline tests/chainer_tests/links_tests/connection_tests/test_convolution_nd.py]: Ditto.

File: tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py [hunk 5]
```diff
@@ -5,137 +5,101 @@
...
 
 
 @testing.parameterize(*testing.product({
     'x_dtype': [numpy.float16, numpy.float32, numpy.float64],
     'W_dtype': [numpy.float16, numpy.float32, numpy.float64],
 }))
-class TestConvolution2D(unittest.TestCase):
+@testing.inject_backend_tests(
+    ['test_forward', 'test_backward', 'test_double_backward',
+     'test_pickling'],
```
Comment (niboshi) [inline tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py]: The first argument can be `None`?

File: tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py [hunk 6]
```diff
@@ -151,74 +115,77 @@ def test_col2im_consistency(self):
...
-    def test_forward_consistency_im2col(self):
-        with chainer.using_config('use_cudnn', 'never'):
-            self.check_forward_consistency()
+@testing.parameterize(*testing.product({
+    'conv_args': [((None, 2, 3, 2, 1), {}),
+                  ((2, 3), {'stride': 2, 'pad': 1})],
+}))
+@testing.inject_backend_tests(
+    ['test_forward', 'test_backward', 'test_double_backward',
+     'test_pickling'],
```
Comment (niboshi) [inline tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py]: Ditto.

File: tests/chainer_tests/links_tests/connection_tests/test_convolution_nd.py [hunk 7]
```diff
@@ -26,37 +23,112 @@
     'in_channels': [4, None, 'omit'],
     'groups': [1, 2],
 })))
-class TestConvolutionND(unittest.TestCase):
+@testing.inject_backend_tests(
+    ['test_forward', 'test_backward', 'test_double_backward',
+     'test_pickling'],
```
Comment (niboshi) [inline tests/chainer_tests/links_tests/connection_tests/test_convolution_nd.py]: Ditto.

File: tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py (hunk 2 above)
Comment (emcastillo) [reply tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py]: Yes, I added it in the first version of the test, it is corrected now.
Thanks!

File: tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py (hunk 5 above)
Comment (emcastillo) [reply tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py]: Done, thanks!

**Comment by emcastillo** [comment]
Review comments first round addressed.
Thanks!

File: tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py [hunk 8]
```diff
@@ -5,137 +5,97 @@
...
+        return x,
 
-    @attr.gpu
-    @condition.retry(3)
-    def test_forward_consistency(self):
-        self.check_forward_consistency()
+    def forward_expected(self, link, inputs):
+        x, = inputs
+        y = link(x).array
+        return y,
```
Comment (niboshi) [inline tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py]: Let's use `F.convolution_2d` here.

File: tests/chainer_tests/links_tests/connection_tests/test_convolution_nd.py [hunk 9]
```diff
@@ -26,123 +23,124 @@
...
+        x = numpy.random.uniform(-1, 1, self.x_shape).astype(self.dtype)
+        return x,
 
-    @attr.cudnn
-    @condition.retry(3)
-    def test_forward_consistency(self):
-        self.check_forward_consistency()
+    def forward_expected(self, link, inputs):
+        x, = inputs
+        y = link(x).array
```
Comment (niboshi) [inline tests/chainer_tests/links_tests/connection_tests/test_convolution_nd.py]: Ditto. Let's use `F.convolution_nd`.

File: tests/chainer_tests/links_tests/connection_tests/test_deconvolution_2d.py [hunk 10]
```diff
@@ -21,74 +15,80 @@ def _pair(x):
...
+                stride=self.stride, pad=self.pad, nobias=self.nobias,
+                dilate=self.dilate, groups=self.groups,
+                initialW=initialW,
+                initial_bias=initial_bias)
+
+        return link
+
+    def forward_expected(self, link, inputs):
+        x, = inputs
+        y = link(x).array
```
Comment (niboshi) [inline tests/chainer_tests/links_tests/connection_tests/test_deconvolution_2d.py]: Ditto.

File: tests/chainer_tests/links_tests/connection_tests/test_deconvolution_nd.py [hunk 11]
```diff
@@ -18,114 +13,102 @@
...
-        with chainer.using_config('use_cudnn', self.use_cudnn):
-            self.check_backward(
-                self.link, cuda.to_gpu(self.x), cuda.to_gpu(self.gy))
+    def generate_inputs(self):
+        x = numpy.random.uniform(-1, 1, self.x_shape).astype(self.dtype)
+        return x,
+
+    def forward_expected(self, link, inputs):
+        x, = inputs
+        y = link(x).array
```
Comment (niboshi) [inline tests/chainer_tests/links_tests/connection_tests/test_deconvolution_nd.py]: Ditto.

File: tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py [hunk 12]
```diff
@@ -151,74 +111,73 @@ def test_col2im_consistency(self):
...
+        'use_cuda': [True],
+        'use_cudnn': ['never', 'always'],
+        'cuda_device': [0, 1],
+    })
+    + [
+        {'use_chainerx': True, 'chainerx_device': 'native:0'},
+        {'use_chainerx': True, 'chainerx_device': 'cuda:0'},
+        {'use_chainerx': True, 'chainerx_device': 'cuda:1'},
+    ])
+class TestConvolution2DParameterShapePlaceholder(testing.LinkTestCase):
```
Comment (niboshi) [inline tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py]: How about merging this test with `TestConvolution2D`? (not mandatory)
Is there any reason this test exists separately?

Comment (emcastillo) [reply tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py]: I am quite unsure of it, the only difference I see is that it explicitly copies the link parameters after initialization,  besides that they look identical ...

**Comment by emcastillo** [comment]
Using functions now instead of directly calling link for forward expected

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Please check test failures.

**Comment by emcastillo** [comment]
pfci seemed to be a transient-failure, however as we have increased the number of tests by adding configurations that were not present before we most likely have raised the memory requirements.

Do you think is ok to remove some of the tests configurations? Cuda multi-device for example.

Jenkins tests is cudnn issues, the new format does not retry tests on failures as before so most of the failures are unmasked. All should go away with the next patch.

**Comment by niboshi** [comment]
OK, hmm, so how about applying the new fix in this PR?
I think it's the most straightforward way to go.

**Comment by emcastillo** [comment]
To adjust all the tolerances based on cudnn version?
Ok, i will prepare it now.
Thanks!

**Comment by emcastillo** [comment]
Added!

File: tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py [hunk 13]
```diff
@@ -5,137 +5,121 @@
...
-        self.link.to_gpu()
-        with chainer.using_config('use_cudnn', 'never'):
-            self.check_backward(cuda.to_gpu(self.x), cuda.to_gpu(self.gy))
+            self.check_forward_options.update({'atol': 5e-3, 'rtol': 5e-2})
+            self.check_backward_options.update({'atol': 3e-2, 'rtol': 5e-2})
+
+    def before_test(self, test_name):
+        # cuDNN 5 and 5.1 results suffer from precision issues
+        using_old_cudnn = (self.backend_config.xp is cuda.cupy and
+                           self.backend_config.use_cudnn == 'always'
```
Comment (niboshi) [inline tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py]: Put `and` after the new line. Ditto for other places.

File: tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py [hunk 14]
```diff
@@ -17,7 +17,8 @@
     'W_dtype': [numpy.float16, numpy.float32, numpy.float64],
 }))
 @testing.inject_backend_tests(
-    None,
+    ['test_forward', 'test_backward',
+     'test_pickling'],
```
Comment (niboshi) [inline tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py]: Why is this fix needed?

Comment (emcastillo) [reply tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py]: it was running double_backward which the original test didn't support.
Probably a cause of increased memory consumption

Comment (niboshi) [reply tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py]: How about setting `skip_double_backward_test = True`?

Comment (emcastillo) [reply tests/chainer_tests/links_tests/connection_tests/test_convolution_2d.py]: didnt know about that one! 
Much better :D thanks!

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins failures are related to functions (not links), will be pushed in #7864 
and convolutionnd fp16 tolerance needs to be adjusted.
I will adjust it now and rerun jenkins.

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
@emcastillo Please do not self-trigger jenkins.
The PR will be merged by Mergify without any review, if Jenkins passes.

File: tests/chainer_tests/links_tests/connection_tests/test_convolution_nd.py [hunk 15]
```diff
@@ -26,37 +24,125 @@
...
+        self.ksize = (3,) * self.ndim
+        self.stride = (2,) * self.ndim
+        self.pad = (1,) * self.ndim
+
+        self.x_shape = (2, 4) + self.dims
+
+        self.check_backward_options.update({'eps': 1e-2,
+                                            'atol': 1e-3, 'rtol': 1e-3})
+        if self.dtype == numpy.float16:
+            self.check_forward_options.update({'atol': 5e-4, 'rtol': 5e-3})
```
Comment (niboshi) [inline tests/chainer_tests/links_tests/connection_tests/test_convolution_nd.py]: How about 5e-3 and 5e-2 respectively to match other functions?

Comment (emcastillo) [reply tests/chainer_tests/links_tests/connection_tests/test_convolution_nd.py]: Yes, I copied from the respective convolution_nd function.
I adjust it now

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
CI errors are functions test tolerances addressed in #7864

**Comment by niboshi** [comment]
LGTM

**Pull request merged**